File: packages/economic_agents/src/economic_agents/observability/llm_quality.py
"""LLM decision quality analysis and hallucination detection.

Measures the quality of LLM-based decision-making, including reasoning depth,
consistency, and detection of hallucinations or invalid decisions.
"""

import json
import re
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

import numpy as np


@dataclass
class LLMQualityMetrics:
    """Comprehensive LLM decision quality metrics."""

    reasoning_depth: float  # 0-100
    consistency_score: float  # 0-100
    hallucination_count: int
    average_response_length: float
    structured_output_success_rate: float  # 0-100
    timestamp: datetime = field(default_factory=datetime.now)


@dataclass
class Hallucination:
    """Detected hallucination in LLM decision."""

    cycle: int
    type: str  # "resource", "capability", "state", "other"
    description: str
    severity: str  # "low", "medium", "high", "critical"
    decision_text: str
    state_at_time: dict[str, Any]


class LLMDecisionQualityAnalyzer:
    """Analyzes the quality of LLM-based decision-making."""

    def __init__(self, agent_id: str | None = None, log_dir: str | None = None):
        """Initialize LLM quality analyzer.

        Args:
            agent_id: Agent identifier
            log_dir: Directory to save analysis results
        """
        self.agent_id = agent_id
        self.log_dir = Path(log_dir) if log_dir else Path("./logs/observability")
        self.log_dir.mkdir(parents=True, exist_ok=True)

        self.decisions: list[dict[str, Any]] = []

    def load_decisions(self, decisions: list[dict[str, Any]]):
        """Load LLM decision history for analysis.

        Args:
            decisions: List of LLM decisions with prompts, responses, reasoning
        """
        self.decisions = decisions

    def measure_reasoning_depth(self, decision: dict[str, Any]) -> float:
        """Measure how thorough the LLM's reasoning was.

        Args:
            decision: LLM decision record

        Returns:
            Reasoning depth score (0-100)
        """
        reasoning = decision.get("reasoning", "")

        if not reasoning:
            return 0.0

        # Factors that indicate deep reasoning
        scores = []

        # 1. Length (longer reasoning suggests more thought)
        length_score = min(len(reasoning) / 500 * 100, 100)
        scores.append(length_score)

        # 2. Contains numbers/calculations
        if re.search(r"\d+\.?\d*", reasoning):
            scores.append(100)
        else:
            scores.append(50)

        # 3. Contains conditional reasoning ("if", "because", "therefore")
        reasoning_keywords = ["if", "because", "therefore", "since", "thus", "however", "although"]
        keyword_count = sum(1 for word in reasoning_keywords if word in reasoning.lower())
        keyword_score = min(keyword_count / 3 * 100, 100)
        scores.append(keyword_score)

        # 4. Contains multiple considerations
        sentences = reasoning.split(".")
        consideration_score = min(len(sentences) / 5 * 100, 100)
        scores.append(consideration_score)

        # 5. References state variables (shows awareness)
        state_keywords = [
            "balance",
            "compute",
            "hours",
            "task",
            "company",
            "revenue",
            "expense",
        ]
        state_ref_count = sum(1 for word in state_keywords if word in reasoning.lower())
        state_score = min(state_ref_count / 3 * 100, 100)
        scores.append(state_score)

        return float(np.mean(scores))

    def measure_consistency(self, window: int = 10) -> float:
        """Measure how consistent decisions are in similar states.

        Args:
            window: Number of recent decisions to analyze

        Returns:
            Consistency score (0-100)
        """
        if len(self.decisions) < 2:
            return 100.0  # Default to perfect if insufficient data

        recent_decisions = self.decisions[-window:]

        # Group similar states
        state_clusters = self._cluster_similar_states(recent_decisions)

        if not state_clusters:
            return 100.0

        # Measure decision variance within clusters
        consistency_scores = []
        for cluster in state_clusters:
            if len(cluster) < 2:
                continue

            # Extract actions from cluster
            actions = [d.get("action", {}) for d in cluster]

            # Calculate variance for each action dimension
            variances = []
            for key in ["task_work_hours", "company_investment_hours", "rest_hours"]:
                values = [a.get(key, 0.0) for a in actions]
                if len(values) > 1:
                    variance = np.std(values)
                    # Normalize variance to 0-100 scale (lower variance = higher consistency)
                    consistency = max(0, 100 - variance * 20)
                    variances.append(consistency)

            if variances:
                consistency_scores.append(np.mean(variances))

        return np.mean(consistency_scores) if consistency_scores else 100.0

    def calculate_overall_quality(self) -> LLMQualityMetrics:
        """Calculate comprehensive LLM quality metrics.

        Returns:
            LLMQualityMetrics with all quality measurements
        """
        if not self.decisions:
            return LLMQualityMetrics(
                reasoning_depth=0.0,
                consistency_score=0.0,
                hallucination_count=0,
                average_response_length=0.0,
                structured_output_success_rate=0.0,
            )

        # Measure reasoning depth for all decisions
        reasoning_depths = [self.measure_reasoning_depth(d) for d in self.decisions]
        avg_reasoning_depth = np.mean(reasoning_depths) if reasoning_depths else 0.0

        # Measure consistency
        consistency = self.measure_consistency()

        # Count hallucinations
        hallucination_detector = HallucinationDetector()
        hallucination_detector.load_decisions(self.decisions)
        hallucinations = hallucination_detector.detect_all_hallucinations()

        # Calculate response lengths
        response_lengths = [len(d.get("reasoning", "")) for d in self.decisions]
        avg_response_length = np.mean(response_lengths) if response_lengths else 0.0

        # Check structured output success rate
        successful_parses = sum(1 for d in self.decisions if d.get("action") is not None and isinstance(d.get("action"), dict))
        success_rate = (successful_parses / len(self.decisions) * 100) if self.decisions else 0.0

        return LLMQualityMetrics(
            reasoning_depth=float(avg_reasoning_depth),
            consistency_score=float(consistency),
            hallucination_count=len(hallucinations),
            average_response_length=float(avg_response_length),
            structured_output_success_rate=success_rate,
        )

    def _cluster_similar_states(
        self, decisions: list[dict[str, Any]], similarity_threshold: float = 0.2
    ) -> list[list[dict[str, Any]]]:
        """Cluster decisions with similar agent states.

        Args:
            decisions: List of decisions to cluster
            similarity_threshold: Maximum difference to be considered similar

        Returns:
            List of decision clusters
        """
        if not decisions:
            return []

        clusters: list[list[dict[str, Any]]] = []

        for decision in decisions:
            state = decision.get("state", {})

            # Find matching cluster
            matched = False
            for cluster in clusters:
                cluster_state = cluster[0].get("state", {})

                # Calculate state similarity
                if self._states_similar(state, cluster_state, similarity_threshold):
                    cluster.append(decision)
                    matched = True
                    break

            # Create new cluster if no match
            if not matched:
                clusters.append([decision])

        return clusters

    def _states_similar(self, state1: dict[str, Any], state2: dict[str, Any], threshold: float) -> bool:
        """Check if two states are similar.

        Args:
            state1: First state
            state2: Second state
            threshold: Similarity threshold (0-1)

        Returns:
            True if states are similar
        """
        # Compare key state variables
        balance1 = state1.get("balance", 0)
        balance2 = state2.get("balance", 0)

        compute1 = state1.get("compute_hours_remaining", 0)
        compute2 = state2.get("compute_hours_remaining", 0)

        # Normalize differences
        balance_diff = abs(balance1 - balance2) / max(balance1, balance2, 1)
        compute_diff = abs(compute1 - compute2) / max(compute1, compute2, 1)

        # Average difference
        avg_diff = (balance_diff + compute_diff) / 2

        return bool(avg_diff <= threshold)

    def export_quality_analysis(self, output_path: str | None = None):
        """Export quality analysis to JSON file.

        Args:
            output_path: Path to save analysis results
        """
        output_file: Path
        if output_path is None:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            output_file = self.log_dir / f"llm_quality_{timestamp}.json"
        else:
            output_file = Path(output_path)

        # Calculate metrics
        metrics = self.calculate_overall_quality()

        # Compile results
        results = {
            "agent_id": self.agent_id,
            "timestamp": datetime.now().isoformat(),
            "decisions_analyzed": len(self.decisions),
            "quality_metrics": {
                "reasoning_depth": metrics.reasoning_depth,
                "consistency_score": metrics.consistency_score,
                "hallucination_count": metrics.hallucination_count,
                "average_response_length": metrics.average_response_length,
                "structured_output_success_rate": metrics.structured_output_success_rate,
            },
        }

        # Save to file
        output_file.parent.mkdir(parents=True, exist_ok=True)
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2)

        return results


class HallucinationDetector:
    """Detects hallucinations in LLM decisions."""

    def __init__(self):
        """Initialize hallucination detector."""
        self.decisions: list[dict[str, Any]] = []
        self.hallucinations: list[Hallucination] = []

    def load_decisions(self, decisions: list[dict[str, Any]]):
        """Load decision history for hallucination detection.

        Args:
            decisions: List of LLM decisions
        """
        self.decisions = decisions

    def detect_all_hallucinations(self) -> list[Hallucination]:
        """Detect all hallucinations across decision history.

        Returns:
            List of detected hallucinations
        """
        self.hallucinations = []

        for i, decision in enumerate(self.decisions):
            state = decision.get("state", {})
            action = decision.get("action", {})
            reasoning = decision.get("reasoning", "")

            # Check for resource hallucinations
            self._check_resource_hallucinations(i, state, action, reasoning)

            # Check for capability hallucinations
            self._check_capability_hallucinations(i, state, action, reasoning)

            # Check for state hallucinations
            self._check_state_hallucinations(i, state, reasoning)

        return self.hallucinations

    def _check_resource_hallucinations(self, cycle: int, state: dict[str, Any], action: dict[str, Any], reasoning: str):
        """Check for hallucinations about available resources.

        Args:
            cycle: Decision cycle number
            state: Agent state
            action: Decision action
            reasoning: LLM reasoning text
        """
        # Check compute hours
        compute_available = state.get("compute_hours_remaining", 0)
        task_hours = action.get("task_work_hours", 0)
        company_hours = action.get("company_investment_hours", 0)
        total_hours = task_hours + company_hours

        if total_hours > compute_available:
            self.hallucinations.append(
                Hallucination(
                    cycle=cycle,
                    type="resource",
                    description=f"Allocated {total_hours}h compute but only {compute_available}h available",
                    severity="critical",
                    decision_text=reasoning[:200],
                    state_at_time=state,
                )
            )

        # Check balance for investments
        balance = state.get("balance", 0)
        if "invest" in reasoning.lower() and balance < 10:
            self.hallucinations.append(
                Hallucination(
                    cycle=cycle,
                    type="resource",
                    description=f"Mentioned investment with low balance (${balance})",
                    severity="medium",
                    decision_text=reasoning[:200],
                    state_at_time=state,
                )
            )

    def _check_capability_hallucinations(self, cycle: int, state: dict[str, Any], action: dict[str, Any], reasoning: str):
        """Check for hallucinations about agent capabilities.

        Args:
            cycle: Decision cycle number
            state: Agent state
            action: Decision action
            reasoning: LLM reasoning text
        """
        # Check for company-related hallucinations
        has_company = state.get("has_company", False)

        if not has_company:
            # Keywords suggesting company operations when no company exists
            company_keywords = ["company", "team", "employees", "sub-agent", "scaling"]
            for keyword in company_keywords:
                if keyword in reasoning.lower():
                    self.hallucinations.append(
                        Hallucination(
                            cycle=cycle,
                            type="capability",
                            description=f"Mentioned '{keyword}' but no company exists",
                            severity="low",
                            decision_text=reasoning[:200],
                            state_at_time=state,
                        )
                    )
                    break  # Only record once per decision

    def _check_state_hallucinations(self, cycle: int, state: dict[str, Any], reasoning: str):
        """Check for hallucinations about current state.

        Args:
            cycle: Decision cycle number
            state: Agent state
            reasoning: LLM reasoning text
        """
        # Check for incorrect state claims
        balance = state.get("balance", 0)

        if "plenty of money" in reasoning.lower() and balance < 50:
            self.hallucinations.append(
                Hallucination(
                    cycle=cycle,
                    type="state",
                    description=f"Claimed 'plenty of money' but balance is ${balance}",
                    severity="medium",
                    decision_text=reasoning[:200],
                    state_at_time=state,
                )
            )

        if "low on funds" in reasoning.lower() and balance > 100:
            self.hallucinations.append(
                Hallucination(
                    cycle=cycle,
                    type="state",
                    description=f"Claimed 'low on funds' but balance is ${balance}",
                    severity="low",
                    decision_text=reasoning[:200],
                    state_at_time=state,
                )
            )

File: packages/economic_agents/crates/economic-agents-observability/src/llm_quality.rs
//! LLM decision quality analysis and hallucination detection.

use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Comprehensive LLM decision quality metrics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LLMQualityMetrics {
    /// Reasoning depth score (0-100).
    pub reasoning_depth: f64,
    /// Consistency score (0-100).
    pub consistency_score: f64,
    /// Number of hallucinations detected.
    pub hallucination_count: usize,
    /// Average response length.
    pub average_response_length: f64,
    /// Structured output success rate (0-100).
    pub structured_output_success_rate: f64,
    /// Analysis timestamp.
    pub timestamp: DateTime<Utc>,
}

/// Hallucination severity levels.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HallucinationSeverity {
    /// Low severity - minor inconsistency.
    Low,
    /// Medium severity - notable error.
    Medium,
    /// High severity - significant error.
    High,
    /// Critical - could cause harmful decisions.
    Critical,
}

/// Detected hallucination in LLM decision.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hallucination {
    /// Cycle number.
    pub cycle: usize,
    /// Hallucination type.
    pub hallucination_type: String,
    /// Description.
    pub description: String,
    /// Severity level.
    pub severity: HallucinationSeverity,
    /// Relevant decision text.
    pub decision_text: String,
    /// State at time of decision.
    pub state_at_time: HashMap<String, serde_json::Value>,
}

/// LLM decision record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LLMDecisionRecord {
    /// State at decision time.
    pub state: HashMap<String, serde_json::Value>,
    /// Action taken.
    pub action: Option<HashMap<String, serde_json::Value>>,
    /// Reasoning provided.
    pub reasoning: String,
}

/// Analyzes the quality of LLM-based decision-making.
pub struct LLMQualityAnalyzer {
    /// Agent identifier.
    #[allow(dead_code)]
    agent_id: Option<String>,
    /// Decision history.
    decisions: Vec<LLMDecisionRecord>,
}

impl LLMQualityAnalyzer {
    /// Create a new analyzer.
    pub fn new() -> Self {
        Self {
            agent_id: None,
            decisions: Vec::new(),
        }
    }

    /// Create with agent ID.
    pub fn with_agent_id(agent_id: impl Into<String>) -> Self {
        Self {
            agent_id: Some(agent_id.into()),
            decisions: Vec::new(),
        }
    }

    /// Load LLM decision history for analysis.
    pub fn load_decisions(&mut self, decisions: Vec<LLMDecisionRecord>) {
        self.decisions = decisions;
    }

    /// Measure how thorough the LLM's reasoning was.
    pub fn measure_reasoning_depth(&self, decision: &LLMDecisionRecord) -> f64 {
        let reasoning = &decision.reasoning;

        if reasoning.is_empty() {
            return 0.0;
        }

        let mut scores = Vec::new();

        // Length (longer reasoning suggests more thought)
        let length_score = (reasoning.len() as f64 / 500.0 * 100.0).min(100.0);
        scores.push(length_score);

        // Contains numbers/calculations
        let has_numbers = reasoning.chars().any(|c| c.is_ascii_digit());
        scores.push(if has_numbers { 100.0 } else { 50.0 });

        // Contains conditional reasoning
        let reasoning_keywords = [
            "if",
            "because",
            "therefore",
            "since",
            "thus",
            "however",
            "although",
        ];
        let keyword_count = reasoning_keywords
            .iter()
            .filter(|kw| reasoning.to_lowercase().contains(*kw))
            .count();
        let keyword_score = (keyword_count as f64 / 3.0 * 100.0).min(100.0);
        scores.push(keyword_score);

        // Contains multiple considerations
        let sentences = reasoning.split('.').count();
        let consideration_score = (sentences as f64 / 5.0 * 100.0).min(100.0);
        scores.push(consideration_score);

        // References state variables
        let state_keywords = [
            "balance", "compute", "hours", "task", "company", "revenue", "expense",
        ];
        let state_ref_count = state_keywords
            .iter()
            .filter(|kw| reasoning.to_lowercase().contains(*kw))
            .count();
        let state_score = (state_ref_count as f64 / 3.0 * 100.0).min(100.0);
        scores.push(state_score);

        if scores.is_empty() {
            0.0
        } else {
            scores.iter().sum::<f64>() / scores.len() as f64
        }
    }

    /// Measure how consistent decisions are in similar states.
    pub fn measure_consistency(&self, window: usize) -> f64 {
        if self.decisions.len() < 2 {
            return 100.0;
        }

        let recent = if self.decisions.len() > window {
            &self.decisions[self.decisions.len() - window..]
        } else {
            &self.decisions
        };

        // Group similar states
        let clusters = self.cluster_similar_states(recent);

        if clusters.is_empty() {
            return 100.0;
        }

        // Measure decision variance within clusters
        let consistency_scores: Vec<f64> = clusters
            .iter()
            .filter(|cluster| cluster.len() >= 2)
            .map(|cluster| {
                let actions: Vec<_> = cluster.iter().filter_map(|d| d.action.as_ref()).collect();

                if actions.len() < 2 {
                    return 100.0;
                }

                let mut variances = Vec::new();
                for key in ["task_work_hours", "company_investment_hours", "rest_hours"] {
                    let values: Vec<f64> = actions
                        .iter()
                        .filter_map(|a| a.get(key).and_then(|v| v.as_f64()))
                        .collect();

                    if values.len() > 1 {
                        let std_dev = Self::std_dev(&values);
                        let consistency = (100.0 - std_dev * 20.0).max(0.0);
                        variances.push(consistency);
                    }
                }

                if variances.is_empty() {
                    100.0
                } else {
                    variances.iter().sum::<f64>() / variances.len() as f64
                }
            })
            .collect();

        if consistency_scores.is_empty() {
            100.0
        } else {
            consistency_scores.iter().sum::<f64>() / consistency_scores.len() as f64
        }
    }

    /// Calculate comprehensive LLM quality metrics.
    pub fn calculate_overall_quality(&self) -> LLMQualityMetrics {
        if self.decisions.is_empty() {
            return LLMQualityMetrics {
                reasoning_depth: 0.0,
                consistency_score: 0.0,
                hallucination_count: 0,
                average_response_length: 0.0,
                structured_output_success_rate: 0.0,
                timestamp: Utc::now(),
            };
        }

        // Measure reasoning depth for all decisions
        let reasoning_depths: Vec<f64> = self
            .decisions
            .iter()
            .map(|d| self.measure_reasoning_depth(d))
            .collect();
        let avg_reasoning_depth =
            reasoning_depths.iter().sum::<f64>() / reasoning_depths.len() as f64;

        // Measure consistency
        let consistency = self.measure_consistency(10);

        // Detect hallucinations
        let hallucinations = self.detect_hallucinations();

        // Calculate response lengths
        let response_lengths: Vec<usize> =
            self.decisions.iter().map(|d| d.reasoning.len()).collect();
        let avg_response_length =
            response_lengths.iter().sum::<usize>() as f64 / response_lengths.len() as f64;

        // Check structured output success rate
        let successful_parses = self.decisions.iter().filter(|d| d.action.is_some()).count();
        let success_rate = successful_parses as f64 / self.decisions.len() as f64 * 100.0;

        LLMQualityMetrics {
            reasoning_depth: avg_reasoning_depth,
            consistency_score: consistency,
            hallucination_count: hallucinations.len(),
            average_response_length: avg_response_length,
            structured_output_success_rate: success_rate,
            timestamp: Utc::now(),
        }
    }

    /// Detect hallucinations in decisions.
    pub fn detect_hallucinations(&self) -> Vec<Hallucination> {
        let mut hallucinations = Vec::new();

        for (i, decision) in self.decisions.iter().enumerate() {
            // Check for resource hallucinations
            if let Some(h) = self.check_resource_hallucination(i, decision) {
                hallucinations.push(h);
            }

            // Check for capability hallucinations
            if let Some(h) = self.check_capability_hallucination(i, decision) {
                hallucinations.push(h);
            }

            // Check for state hallucinations
            if let Some(h) = self.check_state_hallucination(i, decision) {
                hallucinations.push(h);
            }
        }

        hallucinations
    }

    /// Check for resource-related hallucinations.
    fn check_resource_hallucination(
        &self,
        cycle: usize,
        decision: &LLMDecisionRecord,
    ) -> Option<Hallucination> {
        let reasoning = &decision.reasoning.to_lowercase();
        let balance = decision
            .state
            .get("balance")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);

        // Check if reasoning mentions having lots of money when actually broke
        if balance < 20.0
            && (reasoning.contains("wealthy") || reasoning.contains("abundant resources"))
        {
            return Some(Hallucination {
                cycle,
                hallucination_type: "resource".to_string(),
                description: format!(
                    "Claims abundant resources but balance is only ${:.2}",
                    balance
                ),
                severity: HallucinationSeverity::High,
                decision_text: decision.reasoning.chars().take(200).collect(),
                state_at_time: decision.state.clone(),
            });
        }

        None
    }

    /// Check for capability-related hallucinations.
    fn check_capability_hallucination(
        &self,
        cycle: usize,
        decision: &LLMDecisionRecord,
    ) -> Option<Hallucination> {
        let reasoning = &decision.reasoning.to_lowercase();

        // Check for claims of capabilities that don't exist
        let false_capabilities = [
            "trading stocks",
            "accessing internet",
            "contacting users",
            "modifying code",
            "external api",
        ];

        for cap in false_capabilities {
            if reasoning.contains(cap) {
                return Some(Hallucination {
                    cycle,
                    hallucination_type: "capability".to_string(),
                    description: format!("Claims capability '{}' which is not available", cap),
                    severity: HallucinationSeverity::Medium,
                    decision_text: decision.reasoning.chars().take(200).collect(),
                    state_at_time: decision.state.clone(),
                });
            }
        }

        None
    }

    /// Check for state-related hallucinations.
    fn check_state_hallucination(
        &self,
        cycle: usize,
        decision: &LLMDecisionRecord,
    ) -> Option<Hallucination> {
        let reasoning = &decision.reasoning.to_lowercase();
        let has_company = decision
            .state
            .get("has_company")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        // Check if reasoning claims company when none exists
        if !has_company && reasoning.contains("my company") && reasoning.contains("revenue") {
            return Some(Hallucination {
                cycle,
                hallucination_type: "state".to_string(),
                description: "Claims to have a company but has_company is false".to_string(),
                severity: HallucinationSeverity::Medium,
                decision_text: decision.reasoning.chars().take(200).collect(),
                state_at_time: decision.state.clone(),
            });
        }

        None
    }

    /// Cluster decisions with similar states.
    fn cluster_similar_states<'a>(
        &self,
        decisions: &'a [LLMDecisionRecord],
    ) -> Vec<Vec<&'a LLMDecisionRecord>> {
        // Simple clustering by balance ranges
        let mut low_balance = Vec::new();
        let mut mid_balance = Vec::new();
        let mut high_balance = Vec::new();

        for decision in decisions {
            let balance = decision
                .state
                .get("balance")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);

            if balance < 50.0 {
                low_balance.push(decision);
            } else if balance < 150.0 {
                mid_balance.push(decision);
            } else {
                high_balance.push(decision);
            }
        }

        vec![low_balance, mid_balance, high_balance]
            .into_iter()
            .filter(|c| !c.is_empty())
            .collect()
    }

    /// Calculate standard deviation.
    fn std_dev(values: &[f64]) -> f64 {
        if values.len() < 2 {
            return 0.0;
        }
        let mean = values.iter().sum::<f64>() / values.len() as f64;
        let variance = values.iter().map(|v| (v - mean).powi(2)).sum::<f64>() / values.len() as f64;
        variance.sqrt()
    }
}

impl Default for LLMQualityAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_decision(balance: f64, reasoning: &str) -> LLMDecisionRecord {
        let mut state = HashMap::new();
        state.insert("balance".to_string(), serde_json::json!(balance));
        state.insert("has_company".to_string(), serde_json::json!(false));

        let mut action = HashMap::new();
        action.insert("task_work_hours".to_string(), serde_json::json!(4.0));

        LLMDecisionRecord {
            state,
            action: Some(action),
            reasoning: reasoning.to_string(),
        }
    }

    #[test]
    fn test_reasoning_depth_empty() {
        let analyzer = LLMQualityAnalyzer::new();
        let decision = make_decision(100.0, "");
        let depth = analyzer.measure_reasoning_depth(&decision);
        assert_eq!(depth, 0.0);
    }

    #[test]
    fn test_reasoning_depth_with_content() {
        let analyzer = LLMQualityAnalyzer::new();
        let decision = make_decision(
            100.0,
            "Because my balance is 100, I should focus on tasks. If the balance drops, I'll need to work more. Therefore, I'll allocate 4 hours to tasks.",
        );
        let depth = analyzer.measure_reasoning_depth(&decision);
        assert!(depth > 50.0);
    }

    #[test]
    fn test_hallucination_detection() {
        let mut analyzer = LLMQualityAnalyzer::new();
        let decisions = vec![make_decision(
            10.0,
            "With my abundant resources and wealthy status, I can invest heavily.",
        )];
        analyzer.load_decisions(decisions);

        let hallucinations = analyzer.detect_hallucinations();
        assert!(!hallucinations.is_empty());
        assert_eq!(hallucinations[0].hallucination_type, "resource");
    }

    #[test]
    fn test_overall_quality() {
        let mut analyzer = LLMQualityAnalyzer::new();
        let decisions = vec![
            make_decision(100.0, "Considering my balance of 100, I'll work on tasks."),
            make_decision(110.0, "Balance increased to 110, continuing with tasks."),
        ];
        analyzer.load_decisions(decisions);

        let metrics = analyzer.calculate_overall_quality();
        assert!(metrics.reasoning_depth > 0.0);
        assert_eq!(metrics.structured_output_success_rate, 100.0);
    }
}
